File: 3sem/4lab/Banks/BankAccounts/CreditBankAccount.cs
using Banks.Clocks;
using Banks.Entities;
using Banks.Exceptions;
using Banks.Models.BankConfigurations;

namespace Banks.BankAccounts;

public class CreditBankAccount : IBankAccount
{
    public CreditBankAccount(Bank bank, Client client, IClock clock)
    {
        ArgumentNullException.ThrowIfNull(bank);
        ArgumentNullException.ThrowIfNull(client);
        ArgumentNullException.ThrowIfNull(clock);

        Balance = 0;
        Id = Guid.NewGuid();

        Client = client;
        Configuration = bank.Configuration.CreditConfiguration;

        clock.DayChanged += OnDayChanged;
    }

    public Guid Id { get; }
    public Client Client { get; }
    public CreditBankAccountConfiguration Configuration { get; }
    public decimal Balance { get; private set; }

    public bool CanDeposit(decimal moneyAmount)
    {
        return Client.IsVerified() || moneyAmount <= Configuration.UnverifiedLimit;
    }

    public bool CanWithdraw(decimal moneyAmount)
    {
        if (!Client.IsVerified() && moneyAmount > Configuration.UnverifiedLimit)
            return false;

        return moneyAmount <= Balance + Configuration.CreditLimit;
    }

    public void Deposit(decimal moneyAmount)
    {
        if (moneyAmount < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        Balance += moneyAmount;
    }

    public void Withdraw(decimal moneyAmount)
    {
        if (moneyAmount < 0)
            throw InvalidBankAccountOperationException.OnInvalidMoneyAmount();

        if (!CanWithdraw(moneyAmount))
            InvalidBankAccountOperationException.OnShortageOfMoney(moneyAmount, Balance);

        Balance -= moneyAmount;
    }

    private void OnDayChanged()
    {
        if (Balance < 0 && Balance > -Configuration.CreditLimit)
            Balance -= Configuration.Commission;
    }
}

File: 4sem/1lab/src/main/java/BanksRemake/Domain/BankAccounts/CreditBankAccount.java
package BanksRemake.Domain.BankAccounts;

import BanksRemake.Domain.Clocks.Clock;
import BanksRemake.Domain.Clocks.ClockSubscriber;
import BanksRemake.Domain.Entities.Bank;
import BanksRemake.Domain.Entities.Client;
import BanksRemake.Domain.Exceptions.InvalidBankAccountOperationException;
import BanksRemake.Domain.Models.BankConfigurations.CreditBankAccountConfiguration;

import java.util.UUID;

/**
 * A bank account that allows you to go negative by the amount of the credit limit.
 */
public class CreditBankAccount implements BankAccount, ClockSubscriber {
    private final UUID id;
    private final Client client;
    private final CreditBankAccountConfiguration configuration;
    private double balance;

    public CreditBankAccount(Bank bank, Client client, Clock clock) {
        this.balance = 0;

        this.client = client;
        this.configuration = bank.getConfiguration().creditBankAccountConfiguration();

        id = UUID.randomUUID();

        clock.addSubscriber(this);
    }

    @Override
    public UUID getId() {
        return id;
    }

    @Override
    public double getBalance() {
        return balance;
    }

    @Override
    public Client getClient() {
        return client;
    }

    @Override
    public boolean canDeposit(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        return client.isVerified() || moneyAmount <= configuration.unverifiedLimit();
    }

    @Override
    public boolean canWithdraw(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        if (!client.isVerified() && moneyAmount > configuration.unverifiedLimit()) {
            return false;
        }

        return moneyAmount <= balance + configuration.creditLimit();
    }

    @Override
    public void deposit(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        balance += moneyAmount;
    }

    @Override
    public void withdraw(double moneyAmount) throws InvalidBankAccountOperationException {
        if (moneyAmount < 0) {
            throw InvalidBankAccountOperationException.onInvalidMoneyAmount();
        }

        if (!canWithdraw(moneyAmount)) {
            throw InvalidBankAccountOperationException.onShortageOfMoney(moneyAmount, balance);
        }

        balance -= moneyAmount;
    }

    @Override
    public void reactOnDayChanged() {
        if (balance < 0 && balance > -configuration.creditLimit()) {
            balance -= configuration.commission();
        }
    }

    @Override
    public void reactOnMonthChanged() {
    }
}
